pkg/util/winutil/pdhutil/pdhhelper.go

```diff
@@ -66,12 +70,99 @@ func pdhLookupPerfNameByIndex(ndx int) (string, error) {
 		uintptr(unsafe.Pointer(&len)))
 
 	if r != ERROR_SUCCESS {
-		return name, fmt.Errorf("Error getting perf name for index %d %v", ndx, r)
+		return name, fmt.Errorf("Error getting perf name for index %d (%#x)", ndx, r)
 	}
 	name = windows.UTF16ToString(buf)
 	return name, nil
 }
 
+// TODO: configurable?
+const (
+	PDH_REFRESH_INTERVAL = 60 // in seconds
+)
+// Lock enforces no more than once forceRefresh=false
+// is running concurrently
+var lock_lastPdhRefreshTime sync.Mutex
+// tracks last time a refresh was successful
+// initialize with process init time
+var lastPdhRefreshTime = atomic.NewTime(time.Now())
+func refreshPdhObjectCache(forceRefresh bool) (didrefresh bool, err error) {
+	// Refresh the Windows internal PDH Object cache
+	//
+	// When forceRefresh=false, the cache is refreshed no more frequently
+	// than the PDH_REFRESH_INTERVAL.
+	//
+	// forceRefresh - If true, ignore PDH_REFRESH_INTERVAL and refresh anyway
+	//
+	// returns didrefresh=true if the refresh operation was successful
+	//
+
+	var len uint32
+	var lock_held bool
+
+	// forceRefresh can trigger a refresh without taking the lock
+	// so we track if we need to unlock or not
+	lock_held = false
+
+	// Only refresh at most every PDH_REFRESH_INTERVAL seconds
+	// or when forceRefresh=true
+	if !forceRefresh {
+		// TODO: use TryLock in golang 1.18
+		//       we don't need to block here
+		//       worst case the counter is skipped again until next interval.
+		lock_lastPdhRefreshTime.Lock()
+		lock_held = true
+		timenow := time.Now()
+		// time.Time.Sub() uses a monotonic clock
+		if timenow.Sub(lastPdhRefreshTime.Load()).Seconds() < PDH_REFRESH_INTERVAL {
+			// too soon, skip refresh
+			lock_lastPdhRefreshTime.Unlock()
+			return false, nil
+		}
+	}
+
+	// do the refresh
+	// either forceRefresh=true
+	// or the interval expired and lock is held
+
+	log.Infof("Refreshing performance counters")
+	r, _, _ := procPdhEnumObjects.Call(
+		uintptr(0), // NULL data source, use computer in szMachineName parameter
+		uintptr(0), // NULL use local computer
+		uintptr(0), // NULL don't return output
+		uintptr(unsafe.Pointer(&len)), // output size
+		uintptr(PERF_DETAIL_WIZARD),
+		uintptr(1)) // do refresh
+	if r != PDH_MORE_DATA {
+		e := fmt.Sprintf("Failed to refresh performance counters (%#x)", r)
+		log.Errorf(e)
+		if lock_held {
+			lock_lastPdhRefreshTime.Unlock()
+		}
+		return false, fmt.Errorf(e)
+	}
```

Comment:
The call pattern is very interesting. Usually these type of calls need to be called twice. However this is a special call since we are not interested here in counter names directly but more for its "side effect" internal cache refreshment. I think Python implementation is different (not in our code by in CPython). Was this enough for you to work?

Reply:
Yes this is enough to cause the refresh to occur. You are right that the `win32pdh.EnumObjects` wrapper used in `integrations-core` internally calls the function twice.

Reply:
Yes, most likely they will block each other. However this call we do not expect to happen even once generally and occasionally few more times.

I wonder if it worth to print time taken if it is more that N seconds?

Reply:
Windows internally has a mutex on some Pdh calls so they are reentrant.

However, it is possible for this to cause an error
* pywin32 internally calls `PdhEnumObjects(refresh=True)`, gets allocation size `X`
* Golang or python calls `PdhEnumObjects(refresh=True)`, the refresh increases the number of objects
* pywin32 internally calls `PdhEnumObjects(refresh=False, size=X)`, function fails b/c size `X` is insufficient

I think this same race could also occur when pywin32 or golang calls `PdhEnumObjectItems`, which could cause a check to fail to initialize or to get a value.

This race already exists because of the dedicated python thread running `PdhEnumObjects(refresh=True)` in `integrations-core`.

However, given the the following, I think it is acceptable
* the refresh is infrequent
   * python refresh occurs at most every 60 seconds
   * the golang refresh occurs at most every 60 seconds (and also only on demand)
* the result of the race manifesting is to cause a counter to fail to initialize or get a value, both of which will be retried on next check interval